libyara/modules/pe/pe_utils.c

```diff
@@ -125,48 +125,32 @@ PIMAGE_NT_HEADERS32 pe_get_header(const uint8_t* data, size_t data_size)
 
 PIMAGE_DATA_DIRECTORY pe_get_directory_entry(PE* pe, int entry)
 {
-  PIMAGE_DATA_DIRECTORY directory_start;
-  uint8_t* optional_header_start;
-  uint16_t optional_header_size;
-
-  // We are specifically NOT checking NumberOfRvaAndSizes here because it can
-  // lie. 7ff1bf680c80fd73c0b35084904848b3705480ddeb6d0eff62180bd14cd18570 has
-  // NumberOfRvaAndSizes set to 11 when in fact there is a valid
-  // IMAGE_DIRECTORY_ENTRY_COM_DESCRIPTOR entry (which is more than 11). If we
-  // are overly strict here and only parse entries which are less than
-  // NumberOfRvaAndSizes we run the risk of missing otherwise perfectly valid
-  // files. Instead of being strict we check to make sure the entry is within
-  // the OptionalHeader, since SizeOfOptionalHeader includes the DataDirectory
-  // array.
-
-  // In case someone requests an entry which is, by definition, invalid.
-  if (entry >= IMAGE_NUMBEROF_DIRECTORY_ENTRIES)
+  // Explanation
+  // https://github.com/VirusTotal/yara/issues/1525
+
+  // In Windows, any access to data directories is controlled by the
+  // RtlImageDirectoryEntryToData function. This is one of the first checks
+  // whether the desired directory entry is smaller than NumberOfRvaAndSizes.
+  // If it's not, it means that the requested directory entry is not there.
+
+  //  1. Must NOT check whether NumberOfRvaAndSizes > 0x10
+
+  //  2. Must check for NumberOfRvaAndSizes < DataDirectory except for 0x0E for 32-bit binary
```

Comment:
Minor nit: Change `0x0E` here to `IMAGE_DIRECTORY_ENTRY_COM_DESCRIPTOR` so people won't have to look it up. ;)